content/posts/event-driven-scaling-fission-function-using-keda.md

```diff
@@ -0,0 +1,239 @@
+
+## Preparation for creating Message Queue Trigger
+
+Since we are done with the setup, now we will need following to create message queue trigger:
+
+1. Three Kafka topics
+2. Producer function
+3. Consumer function
+
+### Creating Kafka Topics
+
+Now, create three Kafka topics as mentioned below, replace namespace and cluster with yours.
+
+1. Kafka topic for function invocation
+```
+cat << EOF | kubectl create -n my-kafka-project -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+name: request-topic
+labels:
+    strimzi.io/cluster: "my-cluster"
+spec:
+partitions: 3
+replicas: 2
+EOF
+```
+2. Kafka topic for function response
+```
+cat << EOF | kubectl create -n my-kafka-project -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+name: response-topic
+labels:
+    strimzi.io/cluster: "my-cluster"
+spec:
+partitions: 3
+replicas: 2
+EOF
+```
+3. Kafka topic for error response
+```
+cat << EOF | kubectl create -n my-kafka-project -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+name: error-topic
+labels:
+    strimzi.io/cluster: "my-cluster"
+spec:
+partitions: 3
+replicas: 2
+EOF
+```
+
+### Creating Producer Function
+
+The producer function is a Go program that creates one thousand messages with the timestamp and drops into a topic input. For brevity, all values have been hardcoded in the code itself. Save the following code as producer.go
+
+```
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	sarama "github.com/Shopify/sarama"
+)
+
+// Handler posts a message to Kafka Topic
+func Handler(w http.ResponseWriter, r *http.Request) {
+	brokers := []string{"my-cluster-kafka-brokers.my-kafka-project.svc:9092"}
+	producerConfig := sarama.NewConfig()
+	producerConfig.Producer.RequiredAcks = sarama.WaitForAll
+	producerConfig.Producer.Retry.Max = 100
+	producerConfig.Producer.Retry.Backoff = 100
+	producerConfig.Producer.Return.Successes = true
+	producerConfig.Version = sarama.V1_0_0_0
+	producer, err := sarama.NewSyncProducer(brokers, producerConfig)
+	fmt.Println("Created a new producer ", producer)
+	if err != nil {
+		panic(err)
+	}
+	for msg := 1; msg <= 1000; msg++ {
+		ts := time.Now().Format(time.RFC3339)
+		message := fmt.Sprintf("{\"message_number\": %d, \"time_stamp\": \"%s\"}", msg, ts)
+		_, _, err = producer.SendMessage(&sarama.ProducerMessage{
+			Topic: "request-topic",
+			Value: sarama.StringEncoder(message),
+		})
+
+		if err != nil {
+			w.Write([]byte(fmt.Sprintf("Failed to publish message to topic %s: %v", "request-topic", err)))
+			return
+		}
+	}
+	w.Write([]byte("Successfully sent to request-topic"))
+}
+```
+
+Let’s create the environment and function
+```
+$ fission environment create --name go --image fission/go-env-1.12:1.10.0 --builder fission/go-builder-1.12:1.10.0
```

Comment:
Would this work with go versions higher than 1.12? Just add a note if not - but my guess is that any version should work?

Reply:
Yes, this would work with higher than 1.12 as well.